agdk/agdktunnel/app/src/main/cpp/vibration_manager.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "vibration_manager.hpp"
+#include "android/content/context.hpp"
+#include "android/os/vibrator_manager.hpp"
+#include <android/api-level.h>
+#include <gni/gni.h>
+#include <jni.h>
+
+// Add a VibrationAttribute to the vibrate call on API 33 or higher
+#define USE_VIBRATION_ATTRIBUTE_API_LEVEL 33
+
+// Use VibrationManager to obtain a vibrator on API 31 and higher
+#define USE_VIBRATION_MANAGER_API_LEVEL 31
+
+// Use the VibrationEffect version of the vibrate call on API 26 or higher
+#define USE_VIBRATION_EFFECT_API_LEVEL 26
+
+// android.os.VibrationAttributes.USAGE_CLASS_FEEDBACK
+#define USAGE_CLASS_FEEDBACK 2
+
+// android.os.VibrationEffect.EFFECT_HEAVY_CLICK
+#define EFFECT_HEAVY_CLICK 5
```

Comment:
Since both constants for android.os.VibrationEffect.EFFECT_* and android.os.VibrationAttributes.USAGE_* are part of the platform APIs, is it possible to define these macros in the wrapper classes? Or do we expect developers to lookup these constants to define the macros in their app/game code?

Reply:
The wrapper doesn't support autogenerating constants unfortunately, which is why I put them directly in the game code with a comment link to their refdoc path.